cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,258 @@ where
             imm,
         } = builder.func.dfg.insts[inst]
         {
-            let size = imm.bits();
+            let size_static = imm.bits();
 
-            return tag_memory_region_static(iter_ptr, tagged_ptr, size, builder);
+            return tag_memory_region_static(
+                iter_ptr,
+                tagged_ptr,
+                size_static,
+                size,
+                builder,
+                environ,
+            );
         }
     }
 
     tag_memory_region_dynamic(iter_ptr, tagged_ptr, size, builder, environ)
 }
 
-/// Tag a memory region with a known size.
-fn tag_memory_region_static(
+/// Tag a memory region which has a static size known at compile-time.
+/// Iterate through `iter_ptr` starting at a `starting_offset`.
+/// Optimizations include:
+/// - Use st2g operation to tag two memory regions at once
+fn tag_memory_region_static_without_loop_unrolling_threshhold(
     iter_ptr: Value,
+    starting_offset: i64,
     tagged_ptr: Value,
     size: i64,
     builder: &mut FunctionBuilder,
-) -> WasmResult<()> {
+) {
+    let mut i = starting_offset;
+    while size - i >= 32 {
+        let cur_ptr = builder.ins().iadd_imm(iter_ptr, i);
+        builder.ins().arm64_st2g(tagged_ptr, cur_ptr);
+        i += 32;
+    }
+    if i < size {
+        let cur_ptr = builder.ins().iadd_imm(iter_ptr, i);
+        builder.ins().arm64_stg(tagged_ptr, cur_ptr);
+    }
+}
```

Comment:
@martin-fink Is this what you meant in regards to extracting the duplicated code in `tag_memory_region_static` into a seperate lambda/closure? I think a method works better for documentation purposes, and I also encountered some mutable reference problems with `builder` when using a closure.